### PR TITLE
Implement documentation linting.

### PR DESCRIPTION
**Pull request by Elv13**
Many of you saw this evolve over the last year. Here is a version of this PR which seems to be regression free.

The ugly:

 * Sadly abuse of `ldoc` in ways which should not be allowed. However, we have little choices. Upstream is zombie, vendoring would slow down or block our ability to ship the `doc` package in Debian/Ubuntu/Fedora and a full fork would force the distributions to package it, which most wont bother to do and drop AwesomeWM.

The bad:

 * This adds a dozen new `ldoc` tags. Some of them are very annoying to use such as `@noreturn` and `@nagativeallowed`. That's noisy, but they were required for the schema validation.
 * This took me an obscene amount of willpower/time to complete. It was spectacularly boring to finish.
 * I am the typo king, so I probably regressed a bunch of stuff
 * Some unrelated work, like turning the `taglist` and `tasklist` into objects got pulled into this branch due to rebase problems

The Awesome:

 * This literally found over 200 actual erratas in the doc.
 * Many longstanding complains about the doc usability have been solved
 * Entire classes of issues are gone
 * It will be very hard to regress. This now prints warnings when building AwesomeWM and eventually will make the CI fail (but not right now).
 * I removed many rendering restrictions we previously had by pretty much rewritting ldoc from within ldoc.
 * We have a path forward for a dark mode and the doc is now accessible (high contrast mode is working)
 * A lot of groundwork for automatic code generation (such as signals) is now implemented.
 * It is now possible to cross-reference sections such as `@beautiful` variables. This opens the door to remove a lot of existing post-processing hacks, such as the one which generates the current "Appearance" page.
 * It makes the doc much more consistent. A lot of stuff which were previously written as text are now part of the rendering template (such as enum values or callback function prototypes).
 * A lot of HTML/CSS improvements (thanks to @Aire-One and @sclu1034 for the contributions) to make the doc easier to navigate.
 * The landing page (index.html) table of content is no longer a mess.

This PR implements most of @sclu1034 previous reviews, but since they are floating comments on dangling commits, I might have missed some. I am not sure how to best review this. It's huge and will be very hard to rebase.

### CODE REVIEW

**Comment by Elv13** [comment]
Apparently it will need a lot of work to pass the CI. I will take a look at this next week.

**Review by Aire-One — COMMENTED** [review]
Took me multiple session to review all of this. Phew! 

Because it is so large, I think we shouldn't wait too long before merging this PR (and we shouldn't merge anything else in the meantime to prevent rebase hell) or it'll never be merged.

Most of my comment are things we can incrementally fix, so I don't think I have any blocker (except this performance issue on the tasklist page).

File: docs/config.ld
Comment (Elv13) [reply docs/config.ld:2103]: Yes, however I think that would still be weird. I prefer all summaries to be on their own line,

File: docs/config.ld
Comment (Elv13) [reply docs/config.ld]: done

File: docs/config.ld
Comment (Elv13) [reply docs/config.ld]: done

File: docs/config.ld
Comment (Elv13) [reply docs/config.ld]: done

File: docs/config.ld
Comment (Elv13) [reply docs/config.ld:8]: Thanks a lot for that. But we can't remove it without all distros shipping with 0.26.x first. 

As they say, the best time to plant a tree was 30 years ago, the next best time is now. I wish we had taken over ldoc years ago and dropped the stupid useless sandbox. It would have avoided me all the pain to implement this PR... Now `ldoc` is too zombie to get anything merged.

File: tests/examples/wibox/awidget/tasklist/style_shape_focus_border_width.lua
Comment (Elv13) [reply tests/examples/wibox/awidget/tasklist/style_shape_focus_border_width.lua:1]: How so? Works for me

image

Note that the `wibox.container.background` code use a different implementation for the "real" awesome and the web ones. So it is a little off, but that's not really avoidable without switching to `.png`. IDK if this is what you are referring to?

File: lib/awful/widget/tasklist.lua
Comment (Elv13) [reply lib/awful/widget/tasklist.lua:14]: Do you know a way around this? Last time I checked last year that was the only way to get the CSS to be applied to the images.

File: tests/examples/sequences/client/tasklist_custom_filter1.lua
Comment (Elv13) [reply tests/examples/sequences/client/tasklist_custom_filter1.lua:36]: It seems to work fine 

```
luacheck --version
Luacheck: 0.26.1
Lua: PUC-Rio Lua 5.1
Argparse: 0.7.1
LuaFileSystem: 1.7.0
LuaLanes: Not found
```

File: tests/examples/screen/template.lua
Comment (Elv13) [reply tests/examples/screen/template.lua:119]: This is an hack to make post-processing possible. The general idea is that the .SVG are patched with an hacky script (yes, I tried to use libxml2, but the Lua LGI bindings are unusable) to make them compatible with CSS. That, in itself, make the browser high contrast mode work. But more importantly for most people, it opens the door to implement a dark mode for the doc without the image being black on dark grey.

File: lib/wibox/widget/textclock.lua
Comment (Elv13) [reply lib/wibox/widget/textclock.lua:73]: I know, but it makes ldoc explode... `opt` can't have `,`. in them apparently...

File: lib/wibox/widget/init.lua
Comment (Elv13) [reply lib/wibox/widget/init.lua:65]: In the case of `gears.surface`, the dependency loop is rather obvious (a theme can require access to the surface. But a theme might also has some semi legit reason to call `require("wibox.widget")`, like storing a template. While the surface copy of this function will just blow up, this one might have some subtle issues too, so I don't think it hurt to do the lazy-loaded beautiful on both sides.

File: lib/wibox/widget/base.lua
Comment (Elv13) [reply lib/wibox/widget/base.lua:87]: That's the kind of things that happen when you have to repeat the same boring task 798[1] time ;)

[1] Total number of properties which failed the linting. Most of them failed for multiple reasons.

File: lib/naughty/list/notifications.lua
Comment (Elv13) [reply lib/naughty/list/notifications.lua:379]: Right, I think we can remove those `.ldoc` now. This can be done later in another PR. I only  removed the ones with caused some warnings in the linting code.

File: lib/gears/table.lua
Comment (Elv13) [reply lib/gears/table.lua:2]: done

File: lib/awful/widget/watch.lua
Comment (Elv13) [reply lib/awful/widget/watch.lua:22]: Yeah. Can we do this in another PR? This code was already there, I just fixed the formatting.

File: lib/awful/widget/calendar_popup.lua
Comment (Elv13) [reply lib/awful/widget/calendar_popup.lua]: done

File: lib/awful/tooltip.lua
Comment (Elv13) [reply lib/awful/tooltip.lua:399]: I am not sure. The problem here is that it's not really an enum, it's just some entries in a table. So neither `@propertyvalue` or `@tablerowkey` really work. I am not sure if we should overload the meaning of `@propertyvalue`. This might cause problems later on if we start using the schema to generate auto-completion code for those Lua LSP modules.

What I could do is to add a new tag for the 4-5 places where this problem applies?

File: lib/awful/hotkeys_popup/widget.lua
Comment (Elv13) [reply lib/awful/hotkeys_popup/widget.lua:168]: The new linting code caught this. This is why I wanted to invest the time in this PR. It's so much harder to forget things now.

File: docs/ldoc.ltp
Comment (Elv13) [reply docs/ldoc.ltp:421]: Yeah, probably. Can this wait until a follow-up PR?

File: docs/common/widget.ldoc
Comment (Elv13) [reply docs/common/widget.ldoc]: Indeed

File: docs/common/client_theme.ldoc
Comment (Elv13) [reply docs/common/client_theme.ldoc:449]: copy paste... Fixed.

File: docs/common/widget.ldoc
Comment (Elv13) [reply docs/common/widget.ldoc:43]: I guess with a few 16k displays side by side, it's now possible. I think that's the largest you can make the wibox without running into some X11 limit, but I didn't verify. If you think this should be removed, I have no strong opinion.

**Comment by Elv13** [comment]
All tests now pass. @sclu1034 What's your take on this? Do you want to do another review or do you think we should let @Aire-One comments settle down then merge and iterate?

**Comment by sclu1034** [comment]
> @sclu1034 What's your take on this?

My honest take?
As someone who likes tinkering and exploring stuff, the idea of linting documentation is fun. But in the context of a big project with way too little dev time available, trying to do it with LDoc is pure insanity, and I'm fully convinced that it will come down in flames eventually.

> As they say, the best time to plant a tree was 30 years ago, the next best time is now. I wish we had taken over ldoc years ago and dropped the stupid useless sandbox. It would have avoided me all the pain to implement this PR... Now ldoc is too zombie to get anything merged.

I believe you're making the same mistake again.

For a good review, I would have to properly understand those 2500 lines of LDoc "configuration", but I honestly have no motivation to do that (so big thanks to @Aire-One to have done it). If you have small bits you want to point me to specifically, I might do those.

So, if you really want to go through with this, then yes, just do Aire-One's reviews.

File: docs/common/widget.ldoc
Comment (Aire-One) [reply docs/common/widget.ldoc:43]: I have a preference to remove it, because I don't think it brings any relevant information.

File: lib/wibox/widget/textclock.lua
Comment (Aire-One) [reply lib/wibox/widget/textclock.lua:73]: Hm, with a quick search I found https://github.com/lunarmodules/LDoc/issues/165

> [...] I have also noticed that an option such as opt="a,b" is only recognised as "a.

> [...] As for [this] point, I'd consider that a bug ;) So let's keep tracking!

ahah :sob:

File: lib/awful/widget/tasklist.lua
Comment (Aire-One) [reply lib/awful/widget/tasklist.lua:14]: No solution, only complaints here. I guess we should open an issue about it and take some time to think about it.

Maybe a "quick win" can be to hide the examples generated images behind a "show image"-button. So the user has to individually click each button to load an image, but that fix conflict with the idea of having these images here...

File: tests/examples/wibox/awidget/tasklist/style_shape_focus_border_width.lua
Comment (Aire-One) [reply tests/examples/wibox/awidget/tasklist/style_shape_focus_border_width.lua:1]: Hmm weird, I tried to play a bit with them locally to check it wasn't an issue on my side, and never succeed at generating them with the correct rendering... I'll need to try again :shrug:

**Review by Aire-One — DISMISSED** [review]
Let's merge this and iterate!

Thank you, @Elv13, for the insane amount of work

**Review by Aire-One — APPROVED** [review]

**Comment by Elv13** [comment]
Sorry for the long post.

> But in the context of a big project with way too little dev time available, trying to do it with LDoc is pure insanity, and I'm fully convinced that it will come down in flames eventually.

I actually strongly agree here, and put it in "the ugly" in the PR intro for a reason. It was actually very demoralizing to write. The problem here is that the alternatives are limited. Doing nothing means the current problem remain unfixed (forever). Rewriting the entire doc with a non-luadoc/ldoc/javadoc/doxygen compatible paradigm would be too much effort (which nobody will do for free, most importantly, *I* wont do that) and would actually piss off a lot of users who already use the current doc. Starting from scratch, delete all the effort which went into the current doc (typing, hints, lifecycle, examples) then iterate will 100% make the user experience worst for long enough to kill the project. Forking ldoc or vendoring it means getting kicked out of a lot of distributions because of either policy violations or nobody willing to upgrade the packages (which is why I push back against new dependencies, v3.4 stayed in Debian for like 8 years because if this kind of issues). Writing a new Lua documentation system actually requires someone to do it, a someone which doesn't seem to exist. Getting *features* merged into ldoc doesn't seem possible anymore. I tried and got little results. Even if we did, it would require an actual ldoc release and getting the distribution to adopt it (luarock is not an option for the distro packages).

Another alternative me and a few other past contributors played with was to actually have the ldoc template export serialized Lua instead of HTML. If the current implementation goes down in flame, which I agree with you is likely (eventually), this might be the cheapest way forward. It keeps ldoc parser and custom tags for backward compatibility, but replaces most of the config and large template with a CMake based asset rendering pipeline (which we already have due to the doc examples). We can implement the cross-doc references and linting as scripts instead of as a single file monolith. It also allows to use more modern and modular HTML template engines. There's actually code for this already, but it would require a very large amount of work to catch up with the status quo.

Other advantages of that would be that feature like generating API description which is compatible with the existing LSP Lua backends (auto completion, integrated doc in VSCode/Kate/NVim, etc) becomes rather trivial. Generating the schema/dataset for some of these client side documentation search engine javascript libraries also become possible too.

Again, is this the best use of *my* time: no, it isn't. What is: Adding 2k LoC to config.ld and auto-generating the list of changes I had to do to make the entire doc unified and consistent. It implemented 20-ish actual users requests instead of doing the same thing, but in a better way. It helped find about 200 real mistakes (missing returns, incorrect type, documented elements which did not render due to typos/wrong_tag) which years of eyeballs didn't notice.

However, yes, doing it this way is horrific. Eventually the whole linting part can be deleted and we still keep the actual fixes and extra type hints (which doesn't really abuse of ldoc, beside maybe the cross reference code, but that would be a prime target to remove if we more to a doc asset pipeline or a *better* solution than ldoc).

> I'm also not aware of this kind of linting being a big thing anywhere else, and I doubt that's just for the lack of no one having that idea within the last decades of people documenting their software.

That's not ultra common in FLOSS project, but it's pretty widespread in commercial projects. Some languages just enforce it (Go, Python-black) and some other have tooling for it (clang-tidy, clang XML AST output + linter). Then once done in the source code, the doc generation (Doxyen, Javadoc) ingest already linted types and var names. Doc guidelines enforcement tooling exists for these stacks too. Also some issues we have, like using strings for enums, is something C++/C/Java/Go doesn't have to deal with due thanks to having a more comprehensive type system.

You get to understand what I am optimizing for here. Historically, our API wasn't consistent (camelCase/snake_case/nocase or properties/accessor_methods/static_accessor_functions/read_write_chimera_methods or mutually incompatible APIs). Then AwesomeWM <= 3.5.x was mostly undocumented (ie, single line documentation with no type, no arguments, no return value, no properties doc beside a little hardcoded table/list at the bottom), you had to read the code to get anything done. This was fine 15 years ago, when the target audience were more acclimated into operating this way, but that become less and less true as the newer generation are more invested into the mobile and web stack (rather than native open source projects). While before I became co-maintainer, all of my work was on features and module (which had *a lot* of users), nearly everything I did since was building the current doc and replacing non-standard API (or "hard to use") with, still leaky, but more consistent/comprehensive ones (properties, signals, declarative construction, usable by modules). It came at a price, a pretty shitty one, but succeeded in bringing users, keeping them and reboot a working user community (which I am very glad/thankful you guys oversee and grow, that's extra work I don't I personally don't have to time for anymore).

All of that work built, as you correctly point out, massive amount of tech debt (deprecated stuff, more internal complexity, larger API surface to maintain, more tests, lot of copy/pasted examples). It also squeezes a lot of life out of dead(-adjacent) dependencies (Cairo, Pango, LGI, luajit, ldoc, X11, XCB-*, FreeDesktop-specs). Yet, we get new users and some features and some old complains are heard less and less. Are users happy with the doc (before or after this)? No. They never will be. There will always be people who say it's the worst they ever saw or that not having built-in whatever is a show stopped and what not. Always people who complain because it doesn't attempt to teach them Lua. Always some saying it's too little *and* too much doc. Always some who only know how to read Sphinx or Mozilla style doc created by professional technical writers. But yet, with every new doc-wide modernization, we get less of those complaints. So what I am optimizing for here is to lower the effort to do doc-wide changes, fix entire classes of doc issues and keep them solved. It's a tradeoff, it brings maintenance burden in non-core areas of the project, but it lower the cost of adding doc-wide improvements (for me at least, because I know how that tingling system work, at least I currently do...).

So yes, I am inclined into going ahead with this. Future me might hate current me (your arguments are *very* valid), but this 12k line of changes achieve what it tries to achieve, which is improving completeness and consistency of what the user see when they read the doc. It's something which we can ship now and get value now instead of daydreaming of a better world. I genuinely think whatever improves the learning curve is the best use of my time rather than bugs, releases and fancy features.

**Pull request merged**
